File: kraken/src/main/java/glitch/kraken/services/request/TopClipsRequest.java
package glitch.kraken.services.request;

import glitch.api.http.HttpClient;
import glitch.api.http.Routes;
import glitch.kraken.object.enums.ClipPeriod;
import glitch.kraken.object.json.Channel;
import glitch.kraken.object.json.Clip;
import glitch.kraken.object.json.Game;
import glitch.kraken.object.json.collections.Clips;
import glitch.service.AbstractRequest;
import java.util.*;
import java.util.stream.Collectors;
import reactor.core.publisher.Mono;

public class TopClipsRequest extends AbstractRequest<Clip, Clips> {
    private final Set<Locale> language = new LinkedHashSet<>(28);
    private Channel channel;
    private Game game;
    private String cursor;
    private Long limit;
    private ClipPeriod period;
    private Boolean trending;

    public TopClipsRequest(HttpClient httpClient) {
        super(httpClient, Routes.get("/clips/top").newRequest());
    }

    public TopClipsRequest addLanguage(Locale... language) {
        return addLanguage(Arrays.asList(language));
    }

    public TopClipsRequest addLanguage(String... language) {
        return addLanguage(Arrays.stream(language).map(Locale::forLanguageTag).collect(Collectors.toSet()));
    }

    public TopClipsRequest addLanguage(Collection<Locale> language) {
        this.language.addAll(language);
        return this;
    }

    public TopClipsRequest setChannel(Channel channel) {
        this.channel = channel;
        return this;
    }

    public TopClipsRequest setGame(Game game) {
        this.game = game;
        return this;
    }

    public TopClipsRequest setCursor(String cursor) {
        this.cursor = cursor;
        return this;
    }

    public TopClipsRequest setLimit(Long limit) {
        this.limit = limit;
        return this;
    }

    public TopClipsRequest setPeriod(ClipPeriod period) {
        this.period = period;
        return this;
    }

    public TopClipsRequest setTrending(Boolean trending) {
        this.trending = trending;
        return this;
    }

    @Override
    public Mono<Clips> get() {

        if (channel != null) {
            request.queryParam("channel", channel.getUsername());
        } else if (game != null) {
            request.queryParam("game", game.getName());
        }

        if (cursor != null) {
            request.queryParam("cursor", cursor);
        }

        if (!language.isEmpty() && language.size() <= 28) {
            request.queryParam("language", language.stream().map(Locale::getLanguage).collect(Collectors.joining(",")));
        }

        if (limit != null && limit > -1 && limit <= 100) {
            request.queryParam("limit", limit);
        }

        if (period != null) {
            request.queryParam("period", period.name().toLowerCase());
        }

        if (trending != null) {
            request.queryParam("trending", trending);
        }

        return httpClient.exchangeAs(request, Clips.class);
    }
}

File: kraken/src/main/kotlin/io/glitchlib/v5/service/request/TopClipsRequest.kt
package io.glitchlib.v5.service.request

import io.glitchlib.internal.http.AbstractRequest
import io.glitchlib.internal.http.HttpRequest
import io.glitchlib.v5.model.ClipPeriod
import io.glitchlib.v5.model.json.Channel
import io.glitchlib.v5.model.json.Game
import java.util.*

class TopClipsRequest : AbstractRequest {

    private val language = LinkedHashSet<Locale>(28)
    private var channel: Channel? = null
    private var game: Game? = null
    private var cursor: String? = null
    private var limit: Long? = null
    private var period: ClipPeriod? = null
    private var trending: Boolean? = null

    fun addLanguage(vararg language: Locale) = addLanguage(language.toList())
    fun addLanguage(vararg language: String) = addLanguage(language.map(Locale::forLanguageTag))

    fun addLanguage(language: Collection<Locale>) = apply {
        this.language.addAll(language)
    }

    fun channel(channel: Channel) = apply {
        this.channel = channel
    }

    fun game(game: Game) = apply {
        this.game = game
    }

    fun cursor(cursor: String) = apply {
        this.cursor = cursor
    }

    fun limit(limit: Long) = apply {
        this.limit = limit
    }

    fun period(period: ClipPeriod) = apply {
        this.period = period
    }

    fun trending(trending: Boolean) = apply {
        this.trending = trending
    }

    override fun invoke(): HttpRequest.() -> Unit = {
        if (channel != null) {
            addQueryParameters("channel", channel!!.username)
        } else if (game != null) {
            addQueryParameters("game", game!!.name)
        }

        if (cursor != null) {
            addQueryParameters("cursor", cursor!!)
        }

        if (language.isNotEmpty() && language.size <= 28) {
            addQueryParameters("language", language.joinToString(",") { it.language })
        }

        if (limit != null && limit!! > 0 && limit!! <= 100) {
            addQueryParameters("limit", limit.toString())
        }

        if (period != null) {
            addQueryParameters("period", period!!.name.toLowerCase())
        }

        if (trending != null && trending!!) {
            addQueryParameters("trending", trending!!.toString())
        }
    }
}
